aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -1096,6 +1096,128 @@ Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory
   return self;
 }
 
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ covariations ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor cov(const Tensor& self, 
+           const c10::optional<Tensor>& otherOpt,
+           c10::optional<bool> rowvarOpt, 
+           c10::optional<bool> biasOpt, 
+           c10::optional<int64_t> ddofOpt,
+           const c10::optional<Tensor>& fweightsOpt, 
+           const c10::optional<Tensor>& aweightsOpt) {
+  
+  const Tensor& other = c10::value_or_else(otherOpt, [] {return Tensor();});
+  const Tensor& fweights = c10::value_or_else(fweightsOpt, [] {return Tensor();});
+  const Tensor& aweights = c10::value_or_else(aweightsOpt, [] {return Tensor();});
+  auto bias = biasOpt.value_or(false);
+  auto ddof = ddofOpt.value_or(0);
+  auto rowvar = rowvarOpt.value_or(true);
+
+  if(self.size(0) == 0)
+    return at::empty_like(self);  
+  TORCH_CHECK(self.dim() <= 2,
+              "Input1 has more than 2 dimensions." );
+    
+  Tensor X, y;
+  if(self.dim() == 1)
+    X = self.unsqueeze(0);
+  else
+    X = self;
+
+  if(!rowvar && X.size(0) != 1)
+    X = X.t();
+
+  if(other.defined()) {
+    TORCH_CHECK(other.dim() <= 2,
+                "Input2 has more than 2 dimensions." );
+ 
+    if(other.dim() == 1)
+      y = other.unsqueeze(0);
+    else
+      y = other;
+
+    if(!rowvar && y.size(0) != 1)
+      y = y.t();
+
+    std::vector<Tensor> t_list;
+    t_list.push_back(X); t_list.push_back(y);
+    X = at::cat(t_list, 0);
+  }
+
+  if (ddofOpt == c10::nullopt) {
+    ddof = (bias == false) ? 1 : 0;
+  }
+  
+  Tensor w = at::empty(0,self.options());
+  
+  if(fweights.defined()) {
+    TORCH_CHECK((fweights.scalar_type() == at::kLong || fweights.scalar_type() == at::kInt),
+                "fweights must be integer.");
+    TORCH_CHECK(fweights.dim() <= 1,
+                "cannot handle multidimensional fweights.");
+    TORCH_CHECK(fweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+ 
+    for (auto i =0; i<fweights.size(0); i++){
+      if (fweights[i].item<int>() < 0)
+        TORCH_CHECK(false,
+                  "fweights cannot be negative.");
+    }
+    w = at::clone(fweights);
+  }
+
+  if (aweights.defined()) {
+    TORCH_CHECK(aweights.dim() <= 1,
+                "cannot handle multidimensional aweights.");
+    TORCH_CHECK(aweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+    for (auto i =0; i<aweights.size(0); i++){
+      if (aweights[i].item<float>() < 0)
+        TORCH_CHECK(false,
+                  "aweights cannot be negative.");
+    }
+
+    if(w.size(0) == 0)
+      w = at::clone(aweights);
+    else
+      w *= aweights;
+  }
+
+  w = w.to(at::kFloat);
```

Comment:
w is calculated using fweights and aweights.
fweights is strictly integer and aweights are real numbers only, since numpy has used these weights as float64 by default, I tried converting w to float before using it for other operations. 
w cannot be complex, however it can be double. Should I use double instead?

Reply:
You should use the type that matches `self`.